File: src/main/java/al132/alchemistry/compat/jei/category/FissionRecipeCategory.java
package al132.alchemistry.compat.jei.category;

import al132.alchemistry.Alchemistry;
import al132.alchemistry.Ref;
import al132.alchemistry.compat.jei.JEIIntegration;
import al132.alchemistry.recipes.FissionRecipe;
import mezz.jei.api.constants.VanillaTypes;
import mezz.jei.api.gui.IRecipeLayout;
import mezz.jei.api.gui.drawable.IDrawable;
import mezz.jei.api.gui.ingredient.IGuiItemStackGroup;
import mezz.jei.api.helpers.IGuiHelper;
import mezz.jei.api.ingredients.IIngredients;
import mezz.jei.api.recipe.category.IRecipeCategory;
import net.minecraft.client.resources.I18n;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;

import java.util.List;

public class FissionRecipeCategory implements IRecipeCategory<FissionRecipe> {
    private IGuiHelper guiHelper;
    private final static int u = 45;
    private final static int v = 41;

    public FissionRecipeCategory(IGuiHelper guiHelper) {
        this.guiHelper = guiHelper;
    }

    @Override
    public ResourceLocation getUid() {
        return new ResourceLocation(Alchemistry.data.MODID, JEIIntegration.FISSION_CATEGORY);
    }

    @Override
    public Class<? extends FissionRecipe> getRecipeClass() {
        return FissionRecipe.class;
    }

    @Override
    public String getTitle() {
        return I18n.format("alchemistry.jei.fission");
    }

    @Override
    public IDrawable getBackground() {
        return guiHelper.createDrawable(new ResourceLocation(Alchemistry.data.MODID, "textures/gui/fission_gui.png"),
                u, v, 120, 50);
    }

    @Override
    public IDrawable getIcon() {
        return guiHelper.createDrawableIngredient(new ItemStack(Ref.fissionController));
    }

    @Override
    public void setIngredients(FissionRecipe recipe, IIngredients ingredients) {
        ingredients.setInput(VanillaTypes.ITEM, recipe.getInput());
        ingredients.setOutputs(VanillaTypes.ITEM, recipe.getOutputs());
    }

    @Override
    public void setRecipe(IRecipeLayout recipeLayout, FissionRecipe recipe, IIngredients ingredients) {
        IGuiItemStackGroup guiItemStacks = recipeLayout.getItemStacks();

        int x = 48 - u;
        int y = 59 - v;
        guiItemStacks.init(0, true, x, y);
        guiItemStacks.set(0, ingredients.getInputs(VanillaTypes.ITEM).get(0));

        x = 121 - u;
        List<ItemStack> output1 = ingredients.getOutputs(VanillaTypes.ITEM).get(0);
        List<ItemStack> output2 = ingredients.getOutputs(VanillaTypes.ITEM).get(1);
        guiItemStacks.init(1, false, x, y);
        guiItemStacks.set(1, output1);
        x += 18;
        if(!output2.get(0).isEmpty()) {
            guiItemStacks.init(2, false, x, y);
            guiItemStacks.set(2, output2);
        }
    }
}

File: src/main/java/al132/alchemistry/compat/jei/fission/FissionRecipeCategory.kt
package al132.alchemistry.compat.jei.fission

import al132.alchemistry.client.GuiFissionController
import al132.alchemistry.compat.jei.AlchemistryRecipeCategory
import al132.alchemistry.compat.jei.AlchemistryRecipeUID
import al132.alib.utils.Translator
import mezz.jei.api.IGuiHelper
import mezz.jei.api.gui.IRecipeLayout
import mezz.jei.api.ingredients.IIngredients
import mezz.jei.api.ingredients.VanillaTypes

class FissionRecipeCategory(guiHelper: IGuiHelper)
    : AlchemistryRecipeCategory<FissionRecipeWrapper>(guiHelper.createDrawable(guiTexture, u, v, 120, 50),
        "jei.fission_controller.name") {

    override fun getTitle(): String = Translator.translateToLocal("jei.fission_controller.name")

    override fun getUid(): String = AlchemistryRecipeUID.FISSION

    override fun setRecipe(recipeLayout: IRecipeLayout, recipeWrapper: FissionRecipeWrapper, ingredients: IIngredients) {
        val guiItemStacks = recipeLayout.itemStacks

        var x = 48 - u
        var y = 59 - v
        guiItemStacks.init(INPUT_ONE, true, x, y)
        guiItemStacks.set(INPUT_ONE, ingredients.getInputs(VanillaTypes.ITEM)[0])

        x = 121 - u
        val output1 = ingredients.getOutputs(VanillaTypes.ITEM)[0]
        val output2 = ingredients.getOutputs(VanillaTypes.ITEM)[1]
        guiItemStacks.init(OUTPUT_ONE, false, x, y)
        guiItemStacks.set(OUTPUT_ONE, output1)
        x += 18
        guiItemStacks.init(OUTPUT_TWO, false, x, y)
        guiItemStacks.set(OUTPUT_TWO, output2)
    }

    companion object {

        private val INPUT_ONE = 0
        private val OUTPUT_ONE = 1
        private val OUTPUT_TWO = 2

        private val u = 45
        private val v = 41

        private val guiTexture = GuiFissionController.textureLocation
    }
}